src/hotspot/share/memory/metaspace.cpp

```diff
@@ -809,7 +806,30 @@ MetaWord* Metaspace::allocate(ClassLoaderData* loader_data, size_t word_size,
   // Try to allocate metadata.
   MetaWord* result = loader_data->metaspace_non_null()->allocate(word_size, mdtype);
 
+  if (result != NULL) {
+    // Zero initialize.
+    Copy::fill_to_words((HeapWord*)result, word_size, 0);
+
+    log_trace(metaspace)("Metaspace::allocate: type %d return " PTR_FORMAT ".", (int)type, p2i(result));
+  }
+
+  return result;
```

Comment:
Shouldn't we still try to find more memory by calling Universe::heap()->satisfy_failed_metadata_allocation before giving up? Or can that not work if called from the VMThread?

Reply:
It can't work from the VMThread.   Patricio and I were chatting yesterday and he pointed out neither of these VM operations can nest (VM_ChangeBreakpoints and VM_MetaspaceGC) making up names but you get the point.

Reply:
Are you sure?

This means that calls to your new non-TRAPS Metaspace::allocate() fail every time the GC threshold is touched. Since CLMS->allocate() fails not only for "hard" OOMS, when we ran against a hard limit, but also when we hit the GC threshold. satisfy_failed_xxx handles both cases; in the latter case specifically it increases the threshold and retries the allocation.

I guess the next "real" allocate, one done with TRAPS, would then remove the blockage and increase the threshold and life goes on, but meanwhile a number of metaspace allocations could have unnecessarily failed.

It looks like you could solve that by manually calling `ClassLoaderMetaspace::expand_and_allocate` after a failing allocate in your non-TRAPS version; but I'm not completely sure that is right either, since it would mean we miss one GC opportunity.

This stuff feels way too complex.